### PR TITLE
Upgrade gnutls next

### PR DESCRIPTION
**Pull request by yuhaoth**
## Description

Upgrade gnutls-next from 3.6.5 to 3.7.2.
TLS1.3 tests depend on new features in 3.7.2. 
v3.7.2 introduce DISABLE_TLS13_COMPAT_MODE, that can disable compatible middle box. And we will add options for same purpose. That will need test cases for it.

Also, in our development progress, SESSION TICKET and TLS13 COMPATIBLE will not be available at beginning. and the issue target is finish client flow first. we should prepare a well configured server for test

issues: #4417 

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Additional comments
Any additional information that could be of interest

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

## Steps to test or reproduce
Outline the steps to test or reproduce the PR here.

### CODE REVIEW

**Comment by mpg** [comment]
Hi @yuhaoth! A few points:

1. Just for the record, can you briefly explain which features of GnuTLS 3.7 are needed that are missing from 3.6? It's always good to know why we're using a specific version.
2. At a glance, this PR looks fine but it's only a part of what needs to be done: the Jenkins image used by the CI is not generated by the Jenkinsfile in this repo - it's a different one, because it also builds and tests with proprietary toolchains - so we need to update that Jenkinsfile too, and test it before we deploy the new image. I'll share the details with you on slack. (Hopefully in the near future our CI infrastructure is going to be more open.)
3. Since updating the Docker images used by the CI is slightly time-consuming in terms of testing, can you confirm if the current version of `OPENSSL_NEXT` is enough, or if we should upgrade it as well? If that's the case, I'd rather update both of them in one go.

**Comment by mpg** [comment]
>     3\. Since updating the Docker images used by the CI is slightly time-consuming in terms of testing, can you confirm if the current version of `OPENSSL_NEXT` is enough, or if we should upgrade it as well? If that's the case, I'd rather update both of them in one go.

Ok, I had a look at OpenSSL releases and it appears that the current stable version is 1.1.1k which is in the same series as our current `OPENSSL_NEXT` in the CI, which is 1.1.1a. Looking at the [1.1.1 series release notes](https://www.openssl.org/news/openssl-1.1.1-notes.html) I can see that the only changes in this series are fixes - no new feature. So unless we happen to run into one of those bugs, I guess 1.1.1a should be fine.

File: tests/docker/bionic/Dockerfile [hunk 1]
```diff
@@ -137,29 +137,29 @@ RUN cd /tmp \
...
     && /sbin/ldconfig \
     && rm -rf /tmp/nettle*
 
-# Build gnutls next (3.6.5)
+# Build gnutls next (3.7.2)
 RUN cd /tmp \
-    && wget https://www.gnupg.org/ftp/gcrypt/gnutls/v3.6/gnutls-3.6.5.tar.xz -qO- | tar xJ \
-    && cd gnutls-3.6.5 \
-    && ./configure --prefix=/usr/local/gnutls-3.6.5 --exec_prefix=/usr/local/gnutls-3.6.5 \
+    && wget  wget https://www.gnupg.org/ftp/gcrypt/gnutls/v3.7/gnutls-3.7.2.tar.xz -qO- | tar xJ \
```
Comment (mpg) [inline tests/docker/bionic/Dockerfile:152]: Typo: `wget wget`

**Comment by mpg** [comment]
>     1. Just for the record, can you briefly explain which features of GnuTLS 3.7 are needed that are missing from 3.6? It's always good to know why we're using a specific version.

Sorry, that was in the PR: it's about the priority strings `%NO_TICKETS` and `%DISABLE_TLS13_COMPAT_MODE`, and the later is new in 3.7.2. What's lest clear to me, and which I'd like you to clarify for me, is why we need those.

1. According to [the documentation](https://www.gnutls.org/manual/gnutls.html), `%NO_TICKETS` "will prevent the advertizing of the TLS session ticket extension. This is implied by the PFS keyword." Considering all well-behaved clients and servers MUST ignore extensions they don't understand, why do we need this option for GnuTLS not to advertise it?
2. `%DISABLE_TLS13_COMPAT_MODE` "will disable TLS 1.3 middlebox compatibility mode (RFC8446, Appendix D.4) for non-compliant middleboxes." - I'll go read the friendly RFC because I'm not familiar with this compatibility mode yet, but can you explain why we want to disable it for testing? Sorry if that's obvious, it's all a bit new to me. Thanks!

**Review by mpg — APPROVED** [review]
Looks good to me. Apart from reviewing the code, I've tested locally that the Docker image still builds and that the added test passes inside it. I've also tested that the added test is skipped on my machine with an old GnuTLS-next, as expected.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks almost good to me. I have only a few minor suggestions.

File: tests/ssl-opt.sh [hunk 2]
```diff
@@ -405,6 +405,44 @@ requires_gnutls_tls1_3() {
     fi
 }
 
+# check %NO_TICKETS option
+requires_gnutls_next_no_ticket() {
+    requires_gnutls_next
+    if [ "$GNUTLS_NEXT_AVAILABLE" = "NO" ]; then
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply tests/ssl-opt.sh:411]: Can you explain what's happening here? Is something functionally wrong with the current code, or are we talking about simplifications only?

[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:411]: Simplifications only.

[the remark this answers is not in the document]
Comment (yuhaoth) [reply tests/ssl-opt.sh:411]: @ronald-cron-arm , I did not understand your suggestion. It looks you miss something. 
Below is my understand of your comments. But it does not work. `GNUTLS_NO_TICKETS_AVAILABLE` will always be `NO` . 

And I think we should keep similar structure with other function in this file. 

```bash
# Check %NO_TICKETS option
requires_gnutls_next_no_ticket() {
    requires_gnutls_next
    GNUTLS_NO_TICKETS_AVAILABLE="NO"
    if [ -z "${GNUTLS_NO_TICKETS_AVAILABLE:-}" ]; then
        if $GNUTLS_NEXT_CLI --priority-list 2>&1 | grep NO_TICKETS >/dev/null
        then
            GNUTLS_NO_TICKETS_AVAILABLE="YES"
        fi
    fi
    if [ "$GNUTLS_NO_TICKETS_AVAILABLE" = "NO" ]; then
        SKIP_NEXT="YES"
    fi
}
```

[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:411]: You are completely right. Just forget about it and sorry for the noise.

**Review by ronald-cron-arm — APPROVED** [review]

**Review by mpg — APPROVED** [review]
LGTM

**Pull request merged**